rotkehlchen/exchanges/woo.py

```diff
@@ -0,0 +1,537 @@
+                response=response,
+                case='validate_api_key',
+            )
+        return True, ''
+
+    def _api_query(
+            self,
+            endpoint: str,
+            method: Literal['GET', 'POST'] = 'GET',
+            options: Optional[dict[str, Any]] = None,
+    ) -> Response:
+        """
+        Request a  Woo API endpoint (from `endpoint`).
+
+        May Raise:
+        - RemoteError
+        """
+        call_options = options if options else {}
+        request_url = f'{self.base_uri}/{endpoint}'
+        timestamp = str(ts_now_in_ms())
+        parameters = urllib.parse.urlencode(call_options)
+        normalized_content = f'{timestamp}{method}/{endpoint}{parameters}' if endpoint.startswith('v3') else f'{parameters}|{timestamp}'  # noqa: E501
+        signature = hmac.new(
+            self.secret,
+            msg=normalized_content.encode('utf-8'),
+            digestmod=hashlib.sha256,
+        ).hexdigest()
+        self.session.headers.update({
+            'x-api-signature': signature,
+            'x-api-timestamp': timestamp,
+        })
+        log.debug('Woo API request', request_url=request_url, options=options)
+        try:
+            response = self.session.request(
+                method=method,
+                url=request_url,
+                data=call_options if method == 'POST' else {},
+                params=call_options if method == 'GET' else {},
+                headers=self.session.headers,
+            )
+        except requests.exceptions.RequestException as e:
+            raise RemoteError(
+                f'Woo {method} request at {request_url} connection error: {e}.',
+            ) from e
+
+        return response
+
+    @overload
+    def _api_query_paginated(
+            self,
+            start_ts: Timestamp,
+            end_ts: Timestamp,
+            case: Literal['trades'],
+    ) -> list[Trade]:
+        ...
+
+    @overload
+    def _api_query_paginated(
+            self,
+            start_ts: Timestamp,
+            end_ts: Timestamp,
+            case: Literal['asset_movements'],
+    ) -> list[AssetMovement]:
+        ...
+
+    def _api_query_paginated(
+            self,
+            start_ts: Timestamp,
+            end_ts: Timestamp,
+            case: Literal['trades', 'asset_movements'],
+    ) -> Union[list[Trade], list[AssetMovement], list]:
+        """Request a Woo API endpoint paginating via an options attribute."""
+        case_pretty = case.replace('_', ' ')  # human readable case
+        results = []
+        endpoint = 'v1/client/hist_trades' if case == 'trades' else 'v1/asset/history'
+        call_options: dict[str, Any] = {  # they need to be in alphabetic order as stated in their api: https://docs.woo.org/#example  # noqa: E501
+            'end_t': ts_sec_to_ms(end_ts),
+            'fromId': 1,
+            'limit': API_MAX_LIMIT,
+            'start_t': ts_sec_to_ms(start_ts),
+        } if case == 'trades' else {
```

Comment:
Why are you making so many if checks in here man? You realize you can combine all these into one right? You are making the same if check again and again for every variable you build!

This function is badly written again. Only 2 different cases and in each step you have a different if. One if for each variable, one if for deserialization method. Then later down on how to read the response and so on and so forth.

This is just bad code, plain and simple. You need two different separate methods here. It's obvious. There is almost no common logic here. Just a huge function full of ifs.